{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Project 2: Enhancer Classification Problem\n",
    "Nikita Kozlov, 317099\n",
    "\n",
    "The goal of Project 2 is to train a classifier capable of predicting enhancer sequences based on DNA sequence data using the frequency of k-mers.\n",
    "\n"
   ],
   "id": "9b87af7d6dbeb0a5"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-12-07T16:39:26.910565Z",
     "start_time": "2024-12-07T16:39:26.907929Z"
    }
   },
   "source": [
    "# %conda install biopython pybedtools -y\n",
    "# %pip install numpy pandas lightgbm scikit-learn tqdm matplotlib xgboost pyfaidx --upgrade"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Data Preparation\n",
    "\n",
    "Data description:\n",
    "- `experiments.tsv` contains the results of enhancer experiments with following columns:\n",
    "  - curation status: Indicates whether the enhancer’s activity has been experimentally validated (positive or negative).\n",
    "  - coordinate hg38: Contains the genomic coordinates for the hg38 assembly, formatted as chr16:86396481-86397120.\n",
    "  - seq hg38: Provides the DNA sequence for the specified region in the hg38 genome.\n",
    "\n",
    "- `GRCh38.p14.genome.fa` contains the human genome assembly GRCh38 in FASTA format.\n",
    "\n",
    "Positive data: \n",
    "- Records from experiments.tsv.gz where curation `status == ’positive’`.\n",
    "\n",
    "Negative data: \n",
    "- Records where curation `status == ’negative’`.\n",
    "\n",
    "- Random sequences from the entire genome (GRCh38 FASTA file), ensuring:\n",
    "  - No overlap with positive sequences.\n",
    "  - An equal number of negative sequences to positive sequences.\n",
    "  - Sequence lengths matching the lengths of positive sequences.\n",
    "  - No N symbols in the sequences."
   ],
   "id": "c68741597d0fd043"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:39:27.367480Z",
     "start_time": "2024-12-07T16:39:26.913198Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "from pyfaidx import Fasta"
   ],
   "id": "b1320e9e28c9bbb",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:39:27.674584Z",
     "start_time": "2024-12-07T16:39:27.474711Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Load the experiments data\n",
    "experiments = pd.read_csv('data/experiments.tsv', sep='\\t')\n",
    "experiments = experiments[['vista_id', 'curation_status', 'coordinate_hg38', 'seq_hg38']].dropna()\n",
    "experiments"
   ],
   "id": "129b0b4bb76b3495",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     vista_id curation_status           coordinate_hg38  \\\n",
       "0         hs1        positive   chr16:86396481-86397120   \n",
       "1         hs2        negative   chr16:85586489-85588130   \n",
       "2         hs3        negative   chr16:80389446-80390755   \n",
       "3         hs4        positive   chr16:80338700-80339858   \n",
       "4         hs5        negative   chr16:79936010-79937400   \n",
       "...       ...             ...                       ...   \n",
       "4612   mm2322         allelic    chrX:24989747-24991635   \n",
       "4613   mm2322         allelic    chrX:24989747-24991635   \n",
       "4614   mm2323        positive  chr3:157835670-157838142   \n",
       "4615   mm2340        positive   chr14:74203486-74204485   \n",
       "4616   mm2341        positive   chr14:74213757-74214884   \n",
       "\n",
       "                                               seq_hg38  \n",
       "0     AACTGAAGGGACCCCGTTAGCATATAAACAAAAGGTGGGGGGTAGC...  \n",
       "1     GGCCCTGGTATGTTTGTTCTTCCAGGGGCTCCCAGGATGGATCCAG...  \n",
       "2     AAGATTGCCATTTGGGGTGTTTCTTGGGGCTAAGAACCATGAAGAC...  \n",
       "3     CAGAGACAGACAGTGACAGAGACAGATTTTAGAATTTGAACAAAGG...  \n",
       "4     TGACACCCACTATTATCCAGTCCTTGATAAACCTCTTTATTTGTTC...  \n",
       "...                                                 ...  \n",
       "4612  CCATGGGGGGTGGGGGTGGGTGATGAACATGTTTTCTGCTGGGGTA...  \n",
       "4613  CCATGGGGGGTGGGGGTGGGTGATGAACATGTTTTCTGCTGGGGTA...  \n",
       "4614  ttacatgtcttttccttcttgtgtggaattctttgcggattggggt...  \n",
       "4615  CCTTCCCACCCTTTGCCTGGCGCTTTTTCCTCTCGAGCAGCTGGGG...  \n",
       "4616  gcctgaggtttagcgattaaatcactggttcacagtcacctagctg...  \n",
       "\n",
       "[4482 rows x 4 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vista_id</th>\n",
       "      <th>curation_status</th>\n",
       "      <th>coordinate_hg38</th>\n",
       "      <th>seq_hg38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hs1</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr16:86396481-86397120</td>\n",
       "      <td>AACTGAAGGGACCCCGTTAGCATATAAACAAAAGGTGGGGGGTAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hs2</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr16:85586489-85588130</td>\n",
       "      <td>GGCCCTGGTATGTTTGTTCTTCCAGGGGCTCCCAGGATGGATCCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hs3</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr16:80389446-80390755</td>\n",
       "      <td>AAGATTGCCATTTGGGGTGTTTCTTGGGGCTAAGAACCATGAAGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hs4</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr16:80338700-80339858</td>\n",
       "      <td>CAGAGACAGACAGTGACAGAGACAGATTTTAGAATTTGAACAAAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hs5</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr16:79936010-79937400</td>\n",
       "      <td>TGACACCCACTATTATCCAGTCCTTGATAAACCTCTTTATTTGTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4612</th>\n",
       "      <td>mm2322</td>\n",
       "      <td>allelic</td>\n",
       "      <td>chrX:24989747-24991635</td>\n",
       "      <td>CCATGGGGGGTGGGGGTGGGTGATGAACATGTTTTCTGCTGGGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4613</th>\n",
       "      <td>mm2322</td>\n",
       "      <td>allelic</td>\n",
       "      <td>chrX:24989747-24991635</td>\n",
       "      <td>CCATGGGGGGTGGGGGTGGGTGATGAACATGTTTTCTGCTGGGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4614</th>\n",
       "      <td>mm2323</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr3:157835670-157838142</td>\n",
       "      <td>ttacatgtcttttccttcttgtgtggaattctttgcggattggggt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4615</th>\n",
       "      <td>mm2340</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr14:74203486-74204485</td>\n",
       "      <td>CCTTCCCACCCTTTGCCTGGCGCTTTTTCCTCTCGAGCAGCTGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4616</th>\n",
       "      <td>mm2341</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr14:74213757-74214884</td>\n",
       "      <td>gcctgaggtttagcgattaaatcactggttcacagtcacctagctg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4482 rows × 4 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:39:27.705266Z",
     "start_time": "2024-12-07T16:39:27.701719Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Load the genome data\n",
    "genome = Fasta('data/GRCh38.p14.genome.fa')"
   ],
   "id": "960fe02d1e133a3b",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:39:27.737376Z",
     "start_time": "2024-12-07T16:39:27.732224Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Load the positive data\n",
    "positive = experiments[experiments['curation_status'] == 'positive']\n",
    "positive"
   ],
   "id": "c67e4025601d2989",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     vista_id curation_status           coordinate_hg38  \\\n",
       "0         hs1        positive   chr16:86396481-86397120   \n",
       "3         hs4        positive   chr16:80338700-80339858   \n",
       "10       hs12        positive   chr16:78476711-78478047   \n",
       "14       hs16        positive   chr16:72947001-72948646   \n",
       "18       hs20        positive   chr16:72704669-72706250   \n",
       "...       ...             ...                       ...   \n",
       "4609   mm2321        positive    chrX:24999050-25000866   \n",
       "4611   mm2322        positive    chrX:24989747-24991635   \n",
       "4614   mm2323        positive  chr3:157835670-157838142   \n",
       "4615   mm2340        positive   chr14:74203486-74204485   \n",
       "4616   mm2341        positive   chr14:74213757-74214884   \n",
       "\n",
       "                                               seq_hg38  \n",
       "0     AACTGAAGGGACCCCGTTAGCATATAAACAAAAGGTGGGGGGTAGC...  \n",
       "3     CAGAGACAGACAGTGACAGAGACAGATTTTAGAATTTGAACAAAGG...  \n",
       "10    AAGCTAGCTAATTGCTTCTTCAGTTGAAGACCTAAATGAGTTTTAA...  \n",
       "14    GGGCTTCTTGCTATGTCAGCCAATCACGGGGATCCCAAGACGGTAA...  \n",
       "18    aggcagattttgggaggaataaaaggaagcgctagagataaaaaac...  \n",
       "...                                                 ...  \n",
       "4609  CACCATCACCCCTTTCTCCAGCCCTTACTACCTCTTACCTCCAACA...  \n",
       "4611  CCATGGGGGGTGGGGGTGGGTGATGAACATGTTTTCTGCTGGGGTA...  \n",
       "4614  ttacatgtcttttccttcttgtgtggaattctttgcggattggggt...  \n",
       "4615  CCTTCCCACCCTTTGCCTGGCGCTTTTTCCTCTCGAGCAGCTGGGG...  \n",
       "4616  gcctgaggtttagcgattaaatcactggttcacagtcacctagctg...  \n",
       "\n",
       "[2267 rows x 4 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vista_id</th>\n",
       "      <th>curation_status</th>\n",
       "      <th>coordinate_hg38</th>\n",
       "      <th>seq_hg38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hs1</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr16:86396481-86397120</td>\n",
       "      <td>AACTGAAGGGACCCCGTTAGCATATAAACAAAAGGTGGGGGGTAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hs4</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr16:80338700-80339858</td>\n",
       "      <td>CAGAGACAGACAGTGACAGAGACAGATTTTAGAATTTGAACAAAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>hs12</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr16:78476711-78478047</td>\n",
       "      <td>AAGCTAGCTAATTGCTTCTTCAGTTGAAGACCTAAATGAGTTTTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hs16</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr16:72947001-72948646</td>\n",
       "      <td>GGGCTTCTTGCTATGTCAGCCAATCACGGGGATCCCAAGACGGTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>hs20</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr16:72704669-72706250</td>\n",
       "      <td>aggcagattttgggaggaataaaaggaagcgctagagataaaaaac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4609</th>\n",
       "      <td>mm2321</td>\n",
       "      <td>positive</td>\n",
       "      <td>chrX:24999050-25000866</td>\n",
       "      <td>CACCATCACCCCTTTCTCCAGCCCTTACTACCTCTTACCTCCAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4611</th>\n",
       "      <td>mm2322</td>\n",
       "      <td>positive</td>\n",
       "      <td>chrX:24989747-24991635</td>\n",
       "      <td>CCATGGGGGGTGGGGGTGGGTGATGAACATGTTTTCTGCTGGGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4614</th>\n",
       "      <td>mm2323</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr3:157835670-157838142</td>\n",
       "      <td>ttacatgtcttttccttcttgtgtggaattctttgcggattggggt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4615</th>\n",
       "      <td>mm2340</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr14:74203486-74204485</td>\n",
       "      <td>CCTTCCCACCCTTTGCCTGGCGCTTTTTCCTCTCGAGCAGCTGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4616</th>\n",
       "      <td>mm2341</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr14:74213757-74214884</td>\n",
       "      <td>gcctgaggtttagcgattaaatcactggttcacagtcacctagctg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2267 rows × 4 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:39:27.771140Z",
     "start_time": "2024-12-07T16:39:27.766167Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Load the negative data\n",
    "negative = experiments[experiments['curation_status'] == 'negative']\n",
    "negative"
   ],
   "id": "44ab23024a0cf12e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     vista_id curation_status            coordinate_hg38  \\\n",
       "1         hs2        negative    chr16:85586489-85588130   \n",
       "2         hs3        negative    chr16:80389446-80390755   \n",
       "4         hs5        negative    chr16:79936010-79937400   \n",
       "5         hs6        negative    chr16:79916053-79917621   \n",
       "6         hs7        negative    chr16:78992666-78994265   \n",
       "...       ...             ...                        ...   \n",
       "4600   mm2313        negative   chr2:215578659-215580932   \n",
       "4601   mm2314        negative   chr2:215623104-215625068   \n",
       "4602   mm2315        negative   chr2:215665911-215669584   \n",
       "4603   mm2316        negative   chr2:216017447-216018551   \n",
       "4604   mm2317        negative  chr12:113984082-113987735   \n",
       "\n",
       "                                               seq_hg38  \n",
       "1     GGCCCTGGTATGTTTGTTCTTCCAGGGGCTCCCAGGATGGATCCAG...  \n",
       "2     AAGATTGCCATTTGGGGTGTTTCTTGGGGCTAAGAACCATGAAGAC...  \n",
       "4     TGACACCCACTATTATCCAGTCCTTGATAAACCTCTTTATTTGTTC...  \n",
       "5     AGTCACCCAGGTGGTAGTGGGCTGCAGATGCTGTGGGTTTTGTTTC...  \n",
       "6     ACAGAAGCCTCAAGCCTAACCAACAAGAAAGATCACTTCATATGCA...  \n",
       "...                                                 ...  \n",
       "4600  TTGTTAATTAGAAAATGCGTGAGAACCAGGAAAATTTATTATCACT...  \n",
       "4601  atgtaggtagggaccacgtaaggtgctctatacaatcacacagccc...  \n",
       "4602  TTTTTTATTGATTTTGGTGGGAAGATTTCTACCGTTGATTGTATTG...  \n",
       "4603  GTAGCTGAGATCACATTTTGACAATCCTATTTCAGATATCATTCAC...  \n",
       "4604  ctcattttggggatcccctttggccaaaccaaactggaagccagag...  \n",
       "\n",
       "[1913 rows x 4 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vista_id</th>\n",
       "      <th>curation_status</th>\n",
       "      <th>coordinate_hg38</th>\n",
       "      <th>seq_hg38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hs2</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr16:85586489-85588130</td>\n",
       "      <td>GGCCCTGGTATGTTTGTTCTTCCAGGGGCTCCCAGGATGGATCCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hs3</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr16:80389446-80390755</td>\n",
       "      <td>AAGATTGCCATTTGGGGTGTTTCTTGGGGCTAAGAACCATGAAGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hs5</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr16:79936010-79937400</td>\n",
       "      <td>TGACACCCACTATTATCCAGTCCTTGATAAACCTCTTTATTTGTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hs6</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr16:79916053-79917621</td>\n",
       "      <td>AGTCACCCAGGTGGTAGTGGGCTGCAGATGCTGTGGGTTTTGTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hs7</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr16:78992666-78994265</td>\n",
       "      <td>ACAGAAGCCTCAAGCCTAACCAACAAGAAAGATCACTTCATATGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4600</th>\n",
       "      <td>mm2313</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr2:215578659-215580932</td>\n",
       "      <td>TTGTTAATTAGAAAATGCGTGAGAACCAGGAAAATTTATTATCACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4601</th>\n",
       "      <td>mm2314</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr2:215623104-215625068</td>\n",
       "      <td>atgtaggtagggaccacgtaaggtgctctatacaatcacacagccc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4602</th>\n",
       "      <td>mm2315</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr2:215665911-215669584</td>\n",
       "      <td>TTTTTTATTGATTTTGGTGGGAAGATTTCTACCGTTGATTGTATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4603</th>\n",
       "      <td>mm2316</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr2:216017447-216018551</td>\n",
       "      <td>GTAGCTGAGATCACATTTTGACAATCCTATTTCAGATATCATTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4604</th>\n",
       "      <td>mm2317</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr12:113984082-113987735</td>\n",
       "      <td>ctcattttggggatcccctttggccaaaccaaactggaagccagag...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1913 rows × 4 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Data generation for case of using random non-positive data from the genome",
   "id": "5ddb429de9cc1c9d"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:39:28.457264Z",
     "start_time": "2024-12-07T16:39:27.865995Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import os\n",
    "from utils import select_random_sequence\n",
    "from tqdm import tqdm\n",
    "\n",
    "if 'random_negative_sequences.tsv' not in os.listdir('data'):\n",
    "    random_negative_sequences_count = len(negative)\n",
    "    random_negative_sequences_length = int(negative['seq_hg38'].str.len().mean())\n",
    "    random_negative_sequences = []\n",
    "    for _ in tqdm(range(random_negative_sequences_count)):\n",
    "        chromosome, start, end, negative_sequence = select_random_sequence(genome, random_negative_sequences_length, negative)\n",
    "        coordinates = f'{chromosome}:{start}-{end}'\n",
    "        random_negative_sequences.append(['negative', coordinates, negative_sequence])\n",
    "\n",
    "    random_negative_sequences = pd.DataFrame(random_negative_sequences, columns=['curation_status', 'coordinate_hg38', 'seq_hg38'])\n",
    "    random_negative_sequences.to_csv('data/random_negative_sequences.tsv', sep='\\t', index=False)\n",
    "else:\n",
    "    random_negative_sequences = pd.read_csv('data/random_negative_sequences.tsv', sep='\\t')"
   ],
   "id": "91942f242241acea",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Data generation for case of using mixed negative data",
   "id": "5f6cffaa391539e2"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:39:28.509762Z",
     "start_time": "2024-12-07T16:39:28.506301Z"
    }
   },
   "cell_type": "code",
   "source": [
    "mixed_negative_sequences = pd.concat([\n",
    "    negative.sample(len(negative) // 2),\n",
    "    random_negative_sequences.sample(len(random_negative_sequences) // 2)\n",
    "])"
   ],
   "id": "94bc04a6d4fb7b4",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Model Training and Validation\n",
    "\n",
    "- Use at least one classification algorithm (e.g., Random Forest, SVM).\n",
    "- Allocate the last 400 positive and the last 400 negative rows from the `experiments.tsv` file to the test set. Additionally, allocate 400 random sequences to the test set if using a random negative dataset. Do not use these test sequences during model training.\n",
    "- Perform 10-fold cross-validation.\n",
    "- Train the classifier for at least three different values of k (e.g., 3, 4, 5)."
   ],
   "id": "444b7a6807a7a838"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Data from the vista dataset",
   "id": "d0291bfff0b9e56"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:43:21.602765Z",
     "start_time": "2024-12-07T16:42:46.306236Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "from utils import train_pipeline, evaluate_pipeline, transform_vista_dataset_for_classification\n",
    "\n",
    "negative_sequences_configuration = {\n",
    "    'Normal Negative Sequences': negative,\n",
    "    'Random Negative Sequences': random_negative_sequences,\n",
    "    'Mixed Negative Sequences': mixed_negative_sequences,\n",
    "}\n",
    "\n",
    "datasets = {}\n",
    "for train_negative_sequences_name, negative_sequences in negative_sequences_configuration.items():\n",
    "    for k in [3, 4, 5]:\n",
    "        if k not in datasets:\n",
    "            datasets[k] = {}\n",
    "\n",
    "        if train_negative_sequences_name not in datasets[k]:\n",
    "            datasets[k][train_negative_sequences_name] = {}\n",
    "\n",
    "        train_sequences_X, train_sequences_y, test_sequences_X, test_sequences_y = transform_vista_dataset_for_classification(\n",
    "            positive,\n",
    "            negative_sequences,\n",
    "            k=k\n",
    "        )\n",
    "\n",
    "        datasets[k][train_negative_sequences_name] = {\n",
    "            'train_sequences_X': train_sequences_X,\n",
    "            'train_sequences_y': train_sequences_y,\n",
    "            'test_sequences_X': test_sequences_X,\n",
    "            'test_sequences_y': test_sequences_y,\n",
    "        }\n",
    "\n",
    "datasets.keys()"
   ],
   "id": "b2761f04ec250049",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3380/3380 [00:02<00:00, 1560.58it/s]\n",
      "100%|██████████| 800/800 [00:00<00:00, 981.09it/s] \n",
      "100%|██████████| 3380/3380 [00:02<00:00, 1289.73it/s]\n",
      "100%|██████████| 800/800 [00:00<00:00, 896.62it/s]\n",
      "100%|██████████| 3380/3380 [00:04<00:00, 815.39it/s]\n",
      "100%|██████████| 800/800 [00:01<00:00, 649.93it/s]\n",
      "100%|██████████| 3380/3380 [00:02<00:00, 1486.00it/s]\n",
      "100%|██████████| 800/800 [00:00<00:00, 1216.15it/s]\n",
      "100%|██████████| 3380/3380 [00:02<00:00, 1247.34it/s]\n",
      "100%|██████████| 800/800 [00:00<00:00, 1034.73it/s]\n",
      "100%|██████████| 3380/3380 [00:04<00:00, 809.91it/s]\n",
      "100%|██████████| 800/800 [00:01<00:00, 710.98it/s]\n",
      "100%|██████████| 3379/3379 [00:02<00:00, 1501.42it/s]\n",
      "100%|██████████| 800/800 [00:00<00:00, 1216.04it/s]\n",
      "100%|██████████| 3379/3379 [00:02<00:00, 1253.76it/s]\n",
      "100%|██████████| 800/800 [00:00<00:00, 1050.58it/s]\n",
      "100%|██████████| 3379/3379 [00:04<00:00, 807.64it/s]\n",
      "100%|██████████| 800/800 [00:01<00:00, 707.32it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys([3, 4, 5])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-07T16:57:14.777933Z",
     "start_time": "2024-12-07T16:56:26.042232Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import json\n",
    "\n",
    "results = []\n",
    "\n",
    "for k, negative_sequences_configuration in datasets.items():\n",
    "    for train_negative_sequences_name, train_data in negative_sequences_configuration.items():\n",
    "        train_sequences_X = train_data['train_sequences_X']\n",
    "        train_sequences_y = train_data['train_sequences_y']\n",
    "\n",
    "        pipeline = train_pipeline(train_sequences_X, train_sequences_y)\n",
    "\n",
    "        for test_negative_sequences_name, test_data in negative_sequences_configuration.items():\n",
    "            test_sequences_X = test_data['test_sequences_X']\n",
    "            test_sequences_y = test_data['test_sequences_y']\n",
    "\n",
    "            evaluation = evaluate_pipeline(pipeline, test_sequences_X, test_sequences_y)\n",
    "\n",
    "            print(f'k={k}, Trained on {train_negative_sequences_name} vs Tested on {test_negative_sequences_name}')\n",
    "            # print(evaluation)\n",
    "            results.append({\n",
    "                'k': k,\n",
    "                'Train Negative Sequences': train_negative_sequences_name,\n",
    "                'Test Negative Sequences': test_negative_sequences_name,\n",
    "                **evaluation\n",
    "            })\n",
    "\n",
    "with open('results.json', 'w') as f:\n",
    "    json.dump(results, f, indent=2)"
   ],
   "id": "1d3e90e7e25ae703",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k=3, Trained on Normal Negative Sequences vs Tested on Normal Negative Sequences\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  8%|▊         | 67/800 [14:56<2:43:26, 13.38s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k=3, Trained on Normal Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=3, Trained on Normal Negative Sequences vs Tested on Mixed Negative Sequences\n",
      "k=3, Trained on Random Negative Sequences vs Tested on Normal Negative Sequences\n",
      "k=3, Trained on Random Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=3, Trained on Random Negative Sequences vs Tested on Mixed Negative Sequences\n",
      "k=3, Trained on Mixed Negative Sequences vs Tested on Normal Negative Sequences\n",
      "k=3, Trained on Mixed Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=3, Trained on Mixed Negative Sequences vs Tested on Mixed Negative Sequences\n",
      "k=4, Trained on Normal Negative Sequences vs Tested on Normal Negative Sequences\n",
      "k=4, Trained on Normal Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=4, Trained on Normal Negative Sequences vs Tested on Mixed Negative Sequences\n",
      "k=4, Trained on Random Negative Sequences vs Tested on Normal Negative Sequences\n",
      "k=4, Trained on Random Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=4, Trained on Random Negative Sequences vs Tested on Mixed Negative Sequences\n",
      "k=4, Trained on Mixed Negative Sequences vs Tested on Normal Negative Sequences\n",
      "k=4, Trained on Mixed Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=4, Trained on Mixed Negative Sequences vs Tested on Mixed Negative Sequences\n",
      "k=5, Trained on Normal Negative Sequences vs Tested on Normal Negative Sequences\n",
      "k=5, Trained on Normal Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=5, Trained on Normal Negative Sequences vs Tested on Mixed Negative Sequences\n",
      "k=5, Trained on Random Negative Sequences vs Tested on Normal Negative Sequences\n",
      "k=5, Trained on Random Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=5, Trained on Random Negative Sequences vs Tested on Mixed Negative Sequences\n",
      "k=5, Trained on Mixed Negative Sequences vs Tested on Normal Negative Sequences\n",
      "k=5, Trained on Mixed Negative Sequences vs Tested on Random Negative Sequences\n",
      "k=5, Trained on Mixed Negative Sequences vs Tested on Mixed Negative Sequences\n"
     ]
    }
   ],
   "execution_count": 14
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
